docs/development/coding-style-guide.md

```diff
@@ -0,0 +1,262 @@
+# Lokomotive coding style guide
+
+This document describes the coding style that's enforced in the Lokomotive codebase. The guidelines
+in this document should be followed by Lokomotive maintainers and contributors alike. The guide is
+structured in a way which makes it easy to link to a specific guideline, for example from a PR
+comment.
+
+## Guidelines
+
+This section describes the entire set of coding style guidelines which should be followed when
+working on the Lokomotive codebase.
+
+Each "topic" should reside on its own under a heading so that it's easy to unambiguously link to a
+specific guideline during code reviews.
+
+The list of guidelines is meant to evolve over time. If you think a new guideline should be
+enforced, suggest changes to this document in a PR.
+
+### Error string formatting
+
+Error strings are typically *chained*:
+
+```
+Deploying cluster: reading component config: parsing HCL: unexpected character '%'
+```
+
+Remember this fact when constructing error strings and think about how the final string might look
+to the user.
+
+#### Wrapping errors
+
+When "wrapping" an error in another error to provide context, the `%w` formatting verb should be
+used following the [Go 1.13 recommendations](https://blog.golang.org/go1.13-errors) to error
+handling:
+
+`fmt.Errorf("parsing config: %w", err)`
+
+#### Error string text
+
+Don't include words which describe failure in error strings. This includes words such as "failed",
+"error", "could not" etc. It is clear that an error indicates a failure. Instead, provide just the
+*context*, i.e. **what we were doing** when the error occurred (or the erroneous condition if the
+error is the last one in the "chain").
+
+Good:
+
+- `fmt.Errorf("parsing config: %w", err)`
+- `fmt.Errorf("deploying cluster: %w", err)`
+- `fmt.Errorf("reversing time: %w", err)`
+- `fmt.Errorf("space probe is offline")`
+
+Bad:
+
+- `fmt.Errorf("parsing config failed: %w", err)`
+- `fmt.Errorf("error encountered while deploying cluster: %w", err)`
+- `fmt.Errorf("could not reverse time: %w", err)`
```

Comment:
I have mixed feeling with this.

What is the reasoning to prefer those over these? I mean, I understand what was mentioned here: https://github.com/kinvolk/lokomotive/pull/953/files#diff-a4a1697930a32405cd5efdc7aa09b71fR40-R43. But I don't see any real reason to prefer that and when I look at the examples it doesn't seem clearly better.

For example, this line: 
```
 Deploying cluster: reading component config: parsing HCL: unexpected character '%'
```

Seems it will be more clear if it is shown like this:

```
 Deploying cluster: reading component config: parsing HCL config failed: unexpected character '%'
```

If the error clearly says what failed (e.g. "parsing HCL config failed: <err>"), IMHO seems easier to understand rather if it just says what it was doing (e.g. "parsing HCL confing: <err>").

But as I said, I have mixed feeling with this. Sometimes I like the brevity, so I'm unsure.

If no one else has concerns with this, just mark this comment as solved :)

Reply:
Nit-picking, I think using infinitive might help. So, instead of `parsing config` just `parse config`, etc. In fact, in the blog and quickly checking go source code, seems that past and infinitive are quite commonly used.

But maybe this is being too restrictive on how errors should look like with no benefit. As you prefer.

As I said in the previous comment, if no one else has concerns with this, just mark this as solved :)

Reply:
My logic here is the following:

An error indicates to the user that something failed. More precisely, **one specific thing** failed, and caused a chain of dependent things to fail as a result. For example, we couldn't read a file, and therefore we couldn't construct a config, and therefore we couldn't deploy a cluster.

Given the above, including "failed, "couldn't" etc. in *every* piece of the chain is redundant. Consider the following two cases:

```
Error deploying cluster: error reading component config: failed to parse HCL: unexpected character '%' while parsing HCL
```

vs.

```
Error deploying cluster: reading component config: parsing HCL: unexpected character '%'
```

In my opinion there is no need, for example, to say `error reading component config`. It is very clear that we didn't manage to read the config, because if we did there wouldn't be an error...

The pieces of an error string provide **context** to an error as sort of a condensed "story". We don't need to repeat "error", "failed", "this and that broke" etc. to provide context.